src/ray/core_worker/reference_count.cc

```diff
@@ -1003,6 +1009,43 @@ absl::optional<LocalityData> ReferenceCounter::GetLocalityData(
   return locality_data;
 }
 
+void ReferenceCounter::PushToLocationSubscribers(ReferenceTable::iterator it,
+                                                 absl::ReleasableMutexLock &lock) {
+  auto callbacks = it->second.location_subscription_callbacks;
+  it->second.location_subscription_callbacks.clear();
+  auto locations = it->second.locations;
+  auto object_size = it->second.object_size;
+  auto current_version = ++it->second.location_version;
+  // NOTE: We release the lock before running the callbacks.
+  lock.Release();
+  for (const auto callback : callbacks) {
+    callback(locations, object_size, current_version);
+  }
+}
+
+Status ReferenceCounter::GetObjectLocationsAsync(
+    const ObjectID &object_id, uint64_t last_location_version,
+    const LocationSubscriptionCallback &callback) {
+  absl::MutexLock lock(&mutex_);
+  auto it = object_id_refs_.find(object_id);
+  if (it == object_id_refs_.end()) {
+    RAY_LOG(WARNING) << "Tried to register a location subscriber for an object "
+                     << object_id << " that doesn't exist in the reference table";
+    return Status::ObjectNotFound("Object " + object_id.Hex() + " not found");
+  }
+
+  if (last_location_version < it->second.location_version) {
+    // If the last location version is less than the current location version, we
+    // already have location data that the subscriber hasn't seen yet, so we immediately
+    // invoke the callback.
+    callback(it->second.locations, it->second.object_size, it->second.location_version);
```

Comment:
Given that there is only one callback here, and that this callback is going to be immediately posted to the IO event loop, I'm thinking that invoking it here with the lock held is fine.